gson/src/test/java/com/google/gson/JsonArrayAsListSuiteTest.java

```diff
@@ -0,0 +1,68 @@
+package com.google.gson;
+
+import com.google.common.collect.testing.ListTestSuiteBuilder;
+import com.google.common.collect.testing.SampleElements;
+import com.google.common.collect.testing.TestListGenerator;
+import com.google.common.collect.testing.features.CollectionFeature;
+import com.google.common.collect.testing.features.CollectionSize;
+import com.google.common.collect.testing.features.ListFeature;
+import java.util.List;
+import junit.framework.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.AllTests;
+
+/**
+ * Dynamic {@link ListTestSuiteBuilder List test suite} for {@link JsonArray#asList()}. This
+ * complements {@link JsonArrayAsListTest}, which can cover some cases which are not covered here,
+ * e.g. making sure changes in the {@code List} view are visible in the {@code JsonArray}.
+ */
+@RunWith(AllTests.class)
+public class JsonArrayAsListSuiteTest {
+  private static class ListGenerator implements TestListGenerator<JsonElement> {
+    @Override
+    public SampleElements<JsonElement> samples() {
+      return new SampleElements<JsonElement>(
+          JsonNull.INSTANCE,
+          new JsonPrimitive(true),
+          new JsonPrimitive("test"),
+          new JsonArray(),
+          new JsonObject());
+    }
+
+    @Override
+    public JsonElement[] createArray(int length) {
+      return new JsonElement[length];
+    }
+
+    @Override
+    public Iterable<JsonElement> order(List<JsonElement> insertionOrder) {
+      return insertionOrder;
+    }
+
+    @Override
+    public List<JsonElement> create(Object... elements) {
+      JsonArray array = new JsonArray();
+      for (Object element : elements) {
+        array.add((JsonElement) element);
+      }
+      return array.asList();
+    }
+  }
+
+  // Special method recognized by JUnit's `AllTests` runner
+  public static Test suite() {
+    return ListTestSuiteBuilder.using(new ListGenerator())
+        .withFeatures(
+            CollectionSize.ANY,
+            // Note: There is current a Guava bug which causes 'null additions' to not be tested if
+            // 'null queries' is enabled, see https://github.com/google/guava/issues/7401
+            CollectionFeature.ALLOWS_NULL_QUERIES,
+            CollectionFeature.RESTRICTS_ELEMENTS, // List only allows JsonElement
+            CollectionFeature.SUPPORTS_ADD,
+            ListFeature.REMOVE_OPERATIONS,
```

Comment:
We typically use the `GENERAL_PURPOSE` features when applicable. There's such a constant [for `List`](https://github.com/google/guava/blob/d9dbb1594e6d412973e99119bd01ecdca1690baa/guava-testlib/src/com/google/common/collect/testing/features/ListFeature.java#L39) and also [for `Map`](https://github.com/google/guava/blob/d9dbb1594e6d412973e99119bd01ecdca1690baa/guava-testlib/src/com/google/common/collect/testing/features/MapFeature.java#L73). I haven't confirmed whether it applies here, and it's fine to use an explicit list instead even if it _would_ apply, but I wanted to at least point it out.

Reply:
Thanks for the hint! I wasn't completely sure whether they were suitable because `CollectionFeature.GENERAL_PURPOSE` [says](https://github.com/google/guava/blob/3c7c173e9c6ac93f154bfe40876f0c792d849f6e/guava-testlib/src/com/google/common/collect/testing/features/CollectionFeature.java#L86-L91):
> Features supported by general-purpose collections - everything but `RESTRICTS_ELEMENTS`.
>
> See java.util.Collection the definition of general-purpose collections.

And the `List` and `Map` here are not 'general-purpose' in the sense of the `java.util.Collection` definition.